### DISCUSSION TITLE
AWS SAM CLI Formula is incorrect

### DISCUSSION THREAD

**Discussion by jfuss** [Homebrew/discussions#4290] [Everyday usage]
### Output of `brew config`

```shell
Not applicable here
```

### Output of `brew doctor`

```shell
Not applicable here
```

### Description of issue

First off, I am not sure this is the right way to reach out. I was confused this didn't really fit any of the issues and that pointed me to discussions but couldn't find anything to fit this either. So if there is a better place, I am happy to move this to that location

I am one of the maintainer of AWS SAM CLI and was poking around the formula that was added to the Homebrew Core. A couple problems I noticed,
1. This uses pip, which we avoid trying to use in our [tap](https://github.com/aws/homebrew-tap/blob/master/Formula/aws-sam-cli.rb#L20). We fall back if we don't have installers but our installers are much faster and don't mess with python paths on customers machines (a problem we heard about lot before moving to our installers).
2.  The formula in the Core repo removes a file: https://github.com/Homebrew/homebrew-core/blob/master/Formula/aws-sam-cli.rb#L298-L302. This is actually required since these get mounted into a container to help simulate local invokes of AWS Lambda Functions. We do allow running containers of different archs. So removing this completely breaks the CLI.

As a side comment: in creating the AWS SAM CLI Formula in the core repo, it breaks our customers. We have docs that tell customers to `brew tap` then `brew install`. Customers use to this flow and install without providing the tap (which is common), now get the Core repos formula. Which degrades their experience and leaves them with a broken CLI. In the future, maintainers of the Core should reach out before adding. This allows for communication and for people to callout risks to action you (Maintainers of Homebrew) are taking.

**Comment by SMillerDev** [comment]
> don't mess with python paths on customers machines (a problem we heard about lot before moving to our installers).

Homebrew core pip formula all install in a virtual environment so no problem there.

> The formula in the Core repo removes a file

Is there a way to install that file only for the architecture of the host?

> As a side comment: in creating the AWS SAM CLI Formula in the core repo, it breaks our customers. We have docs that tell customers to brew tap then brew install.

That is inconvenient but not really something Homebrew is responsible for.

> maintainers of the Core should reach out before adding.

No, it's an open source project that somebody wanted in Homebrew and they did the work to have it accepted. Upstreams documentation or workflows are of no concern in this.

**Comment by carlocab** [comment]
> 1. This uses pip, which we avoid trying to use in our [tap](https://github.com/aws/homebrew-tap/blob/master/Formula/aws-sam-cli.rb#L20). We fall back if we don't have installers but our installers are much faster and don't mess with python paths on customers machines (a problem we heard about lot before moving to our installers).

We ship binary packages so `brew install aws-sam-cli` never actually runs `pip` for the vast majority of installs. (This will use `pip` on outdated versions of macOS, but that is a small minority of users.) The formula also installs `aws-sam-cli` in a dedicated virtual environment so there is no need to mess with python paths, nor does using the homebrew/core formula mess with the user's python paths.

> 2. The formula in the Core repo removes a file: https://github.com/Homebrew/homebrew-core/blob/master/Formula/aws-sam-cli.rb#L298-L302. This is actually required since these get mounted into a container to help simulate local invokes of AWS Lambda Functions. We do allow running containers of different archs. So removing this completely breaks the CLI.

The CLI should still work if you are running a container of the same arch, so "completely breaks" seems like an unwarranted exaggeration here. Still, this is easy enough to fix; thanks for flagging it. https://github.com/Homebrew/homebrew-core/pull/124616

**Reply by jfuss** [reply]
> We ship binary packages so brew install aws-sam-cli never actually runs pip for the vast majority of installs. (This will use pip on outdated versions of macOS, but that is a small minority of users.) The formula also installs aws-sam-cli in a dedicated virtual environment so there is no need to mess with python paths, nor does using the homebrew/core formula mess with the user's python paths.

Yes you install into an `venv` which is great but how homebrew is suggested to be installed puts homebrew in the front of PATH. Customer who follow the docs, get caught by they `python3` being whatever homebrew resolves too. Yes there are things you can do but the customer perception is "I installed AWS SAM CLI and now my paths are different". It was a common thing we heard from linux users. Maybe things have changed over time, I haven't looked recently but know this was a major reason we have been moving towards our installers over `pip`. 

> The CLI should still work if you are running a container of the same arch, so "completely breaks" seems like an unwarranted exaggeration here.

Maybe an exaggeration, but please keep in mind we are seeing issues now because customers are installing (unknowingly) the wrong formula. Which makes it much hard for my team to be able to investigate issues because what the customer thinks (installing form our tap) is actually not happening. It's also quite common for M1/M2 Macs to run x86 AWS Lambda functions (since that is a common AWS Lambda arch. Our workaround as been, move to the arch you are on but that not helpful.

> Still, this is easy enough to fix; thanks for flagging it. https://github.com/Homebrew/homebrew-core/pull/124616

Thank you. As a follow up, is this the right way to flag things? I was pretty confused last night the best way to reach out.

I am generally curious: What is the value of having this in Homebrew Core? We have a tap and this allows us to control the full cycle of the release. Having it in Core seems fine, but seems to have cause more problems (for us) in the process.

**Reply by SMillerDev** [reply]
> I am generally curious: What is the value of having this in Homebrew Core?

People want it in core, that's enough

> We have a tap and this allows us to control the full cycle of the release.

https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#official-vendor-taps
